embedded_sass.proto

```diff
@@ -0,0 +1,91 @@
+// Copyright 2019 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+syntax = "proto3";
+
+// An inbound request that compiles an entrypoint to CSS.
+message CompileRequest {
+  // This compilation's request id. This is included in messages sent from the
+  // compiler to the host.
+  uint32 id = 1;
+
+  // Possible syntaxes for a Sass stylesheet.
+  enum Syntax {
+    // The CSS-superset `.scss` syntax.
+    SCSS = 0;
+
+    // The indented `.sass` syntax.
+    INDENTED = 1;
+
+    // Plain CSS syntax that doesn't support any special Sass features.
+    CSS = 2;
+  }
+
+  // An input stylesheet provided as plain text, rather than loaded from the
+  // filesystem.
+  message StringInput {
+    // The contents of the stylesheet.
+    string source = 1;
+
+    // The location from which `source` was lodaed. If this is empty, it
+    // indicates that the URL is unknown.
+    string url = 2;
+
+    // The syntax to use to parse `source`.
+    Syntax syntax = 3;
+  }
+
+  // The input stylesheet to parse. This must be set.
+  oneof input {
+    // A stylesheet loaded from its contents.
+    StringInput string = 2;
+
+    // A stylesheet loaded from the given path on the filesystem.
+    string path = 3;
+  }
+
+  // Possible ways to format the CSS output. The compiler is not required to
+  // support all possible options; if the embedder requests an unsupported
+  // style, the compiler should choose the closest supported style.
+  enum OutputStyle {
+    // Each selector and declaration is written on its own line.
+    EXPANDED = 0;
+
+    // The entire stylesheet is written on a single line, with as few characters
+    // as possible.
+    COMPRESSED = 1;
+
+    // CSS rules and declarations are indented to match the nesting of the Sass
+    // source.
+    NESTED = 2;
+
+    // Each CSS rule is written on its own single line, along with all its
+    // declarations.
+    COMPACT = 3;
+  }
+
+  // How to format the CSS output.
+  OutputStyle style = 4;
+
+  // Whether to generate a source map. Note that this will *not* add a source
+  // map comment to the stylesheet; that's up to the embedder or its users.
+  bool source_map = 5;
```

Comment:
Why this behavior while the command-line utility has already the option of embedding?

Reply:
Because the CLI essentially always works with the filesystem, it knows where the Sass source files are read from and where the CSS output is being written. This means it can reliably generate a relative path between the two. When being run in-memory, though, the location of the Sass may not be known (when using StringInput) and the location of the CSS is never known.

It would be possible to add extra metadata to make this work, but I'd rather keep the protocol simple and leave it up to the embedder to understand more about the context in which it's invoked.

Reply:
Makes sense. I wanted to avoid reimplementing this for every client but you have the last word on this.

Reply:
That makes sense too, but there are so many different ways a client might want to link source maps that it'll ultimately be necessary to at least give them the option of handling it all themselves. And at that point, adding support for doing it on the compiler side is just for convenience, which doesn't seem super necessary given how simple the simple case is.